File: app/src/main/java/com/example/game/GameEngine.kt
package com.example.game

import com.example.components.*
import com.example.core.Constants
import com.example.core.GameZone
import com.example.core.Palette
import com.example.core.PowerUpType
import androidx.compose.ui.graphics.toArgb
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue
import androidx.compose.runtime.mutableStateOf

enum class GameScreenState {
    HOME,
    PLAYING,
    PAUSED,
    CRASHING,
    GAME_OVER,
    LEADERBOARD,
    ACHIEVEMENTS,
    SETTINGS
}

class GameEngine(val audioDirector: AudioDirector) {

    var state: GameScreenState by mutableStateOf(GameScreenState.HOME)
        private set

    val player = RickshawPlayer()
    val pools = ObjectPools()
    val spawnDirector = SpawnDirector()
    val difficultyManager = DifficultyManager()

    // Game Session Stats
    var survivalTimeSec: Float = 0f
    var distanceMeters: Float = 0f
    var moneyTaka: Int = 0
    var comboCount: Int = 0
    var comboTimerSec: Float = 0f
    var maxCombo: Int = 0
    var currentZone: GameZone = GameZone.GULISTAN

    // Screen Shake Offset (Integer virtual pixels)
    var shakeOffsetX: Int = 0
    var shakeOffsetY: Int = 0
    private var shakeTimerSec: Float = 0f

    // Crash Slow-mo & Wipe State
    var crashAnimProgress: Float = 0f
    var wipeColumnIndex: Int = -1

    // Active Passenger
    var currentPassenger: PassengerEntity? = null

    // Speed scaling tracking
    var lastThousandCoins: Int = 0
    var obstacleSpawnYTracker: Float = 0f

    // Options
    var useBanglaDigits: Boolean by mutableStateOf(true)
    var showFpsOverlay: Boolean by mutableStateOf(false)

    fun startNewGame() {
        pools.resetAll()
        player.x = Constants.LANE_CENTERS[1] - (Constants.RICKSHAW_WIDTH / 2f)
        player.lane = 1
        player.lives = 3
        player.invincibleTimerSec = 0f
        player.isTurbo = false
        player.isShielded = false
        player.isMagnetActive = false
        player.nearMissCount = 0

        survivalTimeSec = 0f
        distanceMeters = 0f
        moneyTaka = 0
        lastThousandCoins = 0
        obstacleSpawnYTracker = 0f
        spawnDirector.reset()
        comboCount = 0
        comboTimerSec = 0f
        maxCombo = 0
        currentZone = GameZone.GULISTAN
        currentPassenger = null
        crashAnimProgress = 0f
        wipeColumnIndex = -1

        state = GameScreenState.PLAYING
        audioDirector.triggerVoice("abar_suru")
    }

    fun handleSwipe(deltaX: Float) {
        if (state != GameScreenState.PLAYING) return
        if (deltaX > 18f && player.lane < Constants.LANE_COUNT - 1) {
            val newLane = player.lane + 1
            player.startLaneChange(newLane)
            spawnDirector.onPlayerLaneChanged(newLane)
            audioDirector.playSfx("near_miss")
        } else if (deltaX < -18f && player.lane > 0) {
            val newLane = player.lane - 1
            player.startLaneChange(newLane)
            spawnDirector.onPlayerLaneChanged(newLane)
            audioDirector.playSfx("near_miss")
        }
    }

    fun handleTapBell() {
        if (state != GameScreenState.PLAYING) return
        audioDirector.playSfx("bell_ting")

        // Skill Mechanic: Bell causes pedestrians within 40px to dodge 4px earlier!
        for (ped in pools.pedestrianPool) {
            if (ped.isActive && ped.y in (player.y - 40f)..(player.y + 40f)) {
                ped.isDodging = true
                ped.dodgeOffset = if (ped.x < player.x) -6f else 6f
            }
        }
    }

    fun pauseGame() {
        if (state == GameScreenState.PLAYING) {
            state = GameScreenState.PAUSED
        }
    }

    fun resumeGame() {
        if (state == GameScreenState.PAUSED) {
            state = GameScreenState.PLAYING
        }
    }

    fun goToHome() {
        state = GameScreenState.HOME
    }

    fun goToLeaderboard() {
        state = GameScreenState.LEADERBOARD
    }

    fun goToAchievements() {
        state = GameScreenState.ACHIEVEMENTS
    }

    fun goToSettings() {
        state = GameScreenState.SETTINGS
    }

    fun update(dt: Float) {
        if (state != GameScreenState.PLAYING && state != GameScreenState.CRASHING) return

        if (state == GameScreenState.CRASHING) {
            updateCrashSequence(dt)
            return
        }

        survivalTimeSec += dt
        val tier = difficultyManager.getTierForTime(survivalTimeSec)

        // Speed gain for every 1,000 coins collected
        val thousandCoinsCount = moneyTaka / 1000
        if (thousandCoinsCount > lastThousandCoins) {
            lastThousandCoins = thousandCoinsCount
            addReward(0, "⚡ SPEED BOOST!")
            audioDirector.playSfx("turbo_boost")
        }

        val coinSpeedBonus = thousandCoinsCount * 14f
        var currentSpeed = (tier.scrollSpeed + coinSpeedBonus)
        if (player.isTurbo) {
            currentSpeed *= Constants.TURBO_SPEED_MULTIPLIER
        }

        // Advance Distance
        distanceMeters += currentSpeed * dt * Constants.METERS_PER_PIXEL

        // Zone Cycle (Every 900m)
        val zoneIndex = ((distanceMeters / Constants.ZONE_INTERVAL_METERS).toInt()) % GameZone.values().size
        currentZone = GameZone.values()[zoneIndex]

        // Player Update
        player.update(dt)

        // Combo Decay
        if (comboCount > 0) {
            comboTimerSec -= dt
            if (comboTimerSec <= 0f) {
                comboCount = 0
            }
        }

        // Magnet Effect (Pulls coins within 48px toward player)
        if (player.isMagnetActive) {
            for (coin in pools.coinPool) {
                if (coin.isActive && !coin.isCollected) {
                    val dx = player.x + 12f - coin.x
                    val dy = player.y + 16f - coin.y
                    val distSq = dx * dx + dy * dy
                    if (distSq < 48f * 48f) {
                        coin.x += (dx / 4f)
                        coin.y += (dy / 4f)
                    }
                }
            }
        }

        // --- ENTITY SPAWNING ---
        updateSpawning(dt, currentSpeed, tier.maxLanesBlocked)

        // --- MOVEMENT & COLLISION WRAPPER ---
        updateEntitiesAndCollisions(dt, currentSpeed)
    }

    private fun updateSpawning(dt: Float, currentSpeed: Float, maxBlockedLanes: Int) {
        obstacleSpawnYTracker += currentSpeed * dt

        // Obstacle Spawning via Spawn Director
        val spawnDecision = spawnDirector.shouldSpawnObstacleRow(obstacleSpawnYTracker, currentSpeed, maxBlockedLanes)
        if (spawnDecision != null) {
            for (blockedLane in spawnDecision.blockedLanes) {
                val obs = pools.obtainObstacle()
                if (obs != null) {
                    obs.isActive = true
                    obs.x = Constants.LANE_CENTERS[blockedLane] - 8f
                    obs.y = -30f
                    obs.type = spawnDecision.obstacleType
                }
            }
        }

        // Coins Arc Spawning
        if (Math.random() < 0.02) {
            val pattern = spawnDirector.generateCoinPattern(-20f)
            for (info in pattern) {
                val coin = pools.obtainCoin()
                if (coin != null) {
                    coin.isActive = true
                    coin.isCollected = false
                    coin.x = info.x
                    coin.y = info.y
                }
            }
        }

        // Power-Up Spawning
        val pupType = spawnDirector.checkPowerUpSpawn()
        if (pupType != null) {
            val pup = pools.obtainPowerUp()
            if (pup != null) {
                pup.isActive = true
                pup.type = pupType
                val lane = (0 until Constants.LANE_COUNT).random()
                pup.x = Constants.LANE_CENTERS[lane] - 8f
                pup.y = -20f
            }
        }

        // Pedestrians
        if (Math.random() < 0.015) {
            val ped = pools.obtainPedestrian()
            if (ped != null) {
                ped.isActive = true
                ped.isDodging = false
                ped.dodgeOffset = 0f
                val lane = (0 until Constants.LANE_COUNT).random()
                ped.x = Constants.LANE_CENTERS[lane] - 8f
                ped.y = -24f
            }
        }

        // Passengers (Kerb side)
        if (currentPassenger == null && Math.random() < 0.008) {
            val pass = pools.obtainPassenger()
            if (pass != null) {
                pass.isActive = true
                pass.isLeftKerb = Math.random() < 0.5
                pass.x = if (pass.isLeftKerb) 10f else 154f
                pass.y = -24f
                pass.typeIndex = (0..5).random()
                pass.isPickedUp = false
                pass.dropoffTargetMeter = distanceMeters.toInt() + (400..900).random()
            }
        }
    }

    private fun updateEntitiesAndCollisions(dt: Float, currentSpeed: Float) {
        val playerHitbox = player.getHitbox()

        // 1. OBSTACLES
        for (obs in pools.obstaclePool) {
            if (!obs.isActive) continue
            obs.y += currentSpeed * dt

            if (obs.getHitbox().intersects(playerHitbox)) {
                if (player.isTurbo) {
                    // Destroy obstacle
                    obs.isActive = false
                    spawnBurstParticles(obs.x + 8f, obs.y + 8f)
                    addReward(10, "SMASH!")
                } else if (player.isShielded) {
                    // Consume shield
                    player.isShielded = false
                    obs.isActive = false
                    audioDirector.playSfx("near_miss")
                    spawnBurstParticles(obs.x + 8f, obs.y + 8f)
                } else if (player.invincibleTimerSec > 0f) {
                    // Temporarily invincible right after losing a life
                } else {
                    // Lose 1 life on obstacle crash
                    player.lives--
                    obs.isActive = false
                    spawnBurstParticles(obs.x + 8f, obs.y + 8f, count = 16)
                    shakeTimerSec = 0.35f
                    audioDirector.playSfx("crash_boom")

                    if (player.lives > 0) {
                        player.invincibleTimerSec = 1.8f
                        addReward(0, "-1 LIFE! ❤️")
                        audioDirector.triggerVoice("are_baba")
                    } else {
                        triggerCrash()
                        return
                    }
                }
            }

            if (obs.y > Constants.VIRTUAL_HEIGHT + 32f) obs.isActive = false
        }

        // 2. COINS
        for (coin in pools.coinPool) {
            if (!coin.isActive || coin.isCollected) continue
            coin.y += currentSpeed * dt

            if (coin.getHitbox().intersects(playerHitbox)) {
                coin.isCollected = true
                coin.isActive = false
                addReward(20, "৳20")
                audioDirector.playSfx("coin_ching")
            }

            if (coin.y > Constants.VIRTUAL_HEIGHT + 16f) coin.isActive = false
        }

        // 3. POWERUPS
        for (pup in pools.powerUpPool) {
            if (!pup.isActive) continue
            pup.y += currentSpeed * dt

            if (pup.getHitbox().intersects(playerHitbox)) {
                pup.isActive = false
                audioDirector.playSfx("turbo_boost")
                when (pup.type) {
                    PowerUpType.TURBO -> {
                        player.isTurbo = true
                        player.turboTimeSec = pup.type.durationSec
                        audioDirector.triggerVoice("bachao")
                    }
                    PowerUpType.MONEY_MAGNET -> {
                        player.isMagnetActive = true
                        player.magnetTimeSec = pup.type.durationSec
                    }
                    PowerUpType.SHIELD -> {
                        player.isShielded = true
                    }
                    PowerUpType.MONEY_RAIN -> {
                        spawnMoneyRain()
                    }
                }
            }

            if (pup.y > Constants.VIRTUAL_HEIGHT + 16f) pup.isActive = false
        }

        // 4. PASSENGERS
        for (pass in pools.passengerPool) {
            if (!pass.isActive) continue
            pass.y += currentSpeed * dt

            if (!pass.isPickedUp && pass.getHitbox().intersects(playerHitbox)) {
                pass.isPickedUp = true
                pass.isActive = false
                currentPassenger = pass
                addReward(50, "PICKUP!")
                audioDirector.playSfx("pickup_ting")
                audioDirector.triggerVoice("vada_ache")
            }

            if (pass.y > Constants.VIRTUAL_HEIGHT + 24f) pass.isActive = false
        }

        // Passenger Dropoff check
        if (currentPassenger != null && distanceMeters >= currentPassenger!!.dropoffTargetMeter) {
            addReward(200, "DROPOFF!")
            audioDirector.playSfx("dropoff_chime")
            currentPassenger = null
        }

        // 5. PEDESTRIANS & PERSON HITS
        for (ped in pools.pedestrianPool) {
            if (!ped.isActive) continue
            ped.y += currentSpeed * dt

            val pedHitbox = ped.getHitbox()
            if (pedHitbox.intersects(playerHitbox)) {
                // Hitting a person EARNS POINTS!
                ped.isActive = false
                addReward(100, "+100 PTS! 👤")
                audioDirector.playSfx("coin_ching")
                audioDirector.triggerVoice("mama_side_den")
                spawnBurstParticles(ped.x + 8f, ped.y + 12f, count = 12)
            } else {
                // Near miss check (within 6px, no contact)
                val distY = Math.abs(ped.y - player.y)
                if (distY < 6f && !ped.isDodging) {
                    ped.isDodging = true
                    ped.dodgeOffset = if (ped.x < player.x) -8f else 8f
                    player.nearMissCount++
                    addReward(10, "NEAR MISS!")
                    audioDirector.playSfx("near_miss")
                    audioDirector.triggerVoice("mama_side_den")
                }
            }

            if (ped.y > Constants.VIRTUAL_HEIGHT + 24f) ped.isActive = false
        }

        // Update Particles & Floating Text
        updateParticles(dt)
        updateFloatingText(dt)
    }

    private fun addReward(baseTaka: Int, label: String) {
        comboCount++
        comboTimerSec = 3.0f
        if (comboCount > maxCombo) maxCombo = comboCount

        val multiplier = when {
            comboCount >= 20 -> 5
            comboCount >= 10 -> 3
            comboCount >= 5 -> 2
            else -> 1
        }

        val total = baseTaka * multiplier
        moneyTaka += total

        // Floating Text
        val ft = pools.obtainFloatingText()
        if (ft != null) {
            ft.isActive = true
            ft.text = if (multiplier > 1) "$label x$multiplier" else label
            ft.x = player.x
            ft.y = player.y - 10f
            ft.life = 0f
            ft.maxLife = 0.6f
            ft.colorArgb = when (multiplier) {
                5 -> Palette.gold.toArgb()
                3 -> Palette.amber.toArgb()
                2 -> Palette.orange.toArgb()
                else -> Palette.cream.toArgb()
            }
        }
    }

    private fun triggerCrash() {
        player.lives = 0
        state = GameScreenState.CRASHING
        audioDirector.playSfx("crash_boom")
        audioDirector.triggerVoice("are_baba")

        // 24 pixel particles burst
        spawnBurstParticles(player.x + 16f, player.y + 20f, count = 24)

        // Screen Shake ±3px integer decay
        shakeTimerSec = 0.35f
    }

    private fun updateCrashSequence(dt: Float) {
        crashAnimProgress += dt
        if (shakeTimerSec > 0f) {
            shakeTimerSec -= dt
            shakeOffsetX = (-3..3).random()
            shakeOffsetY = (-3..3).random()
        } else {
            shakeOffsetX = 0
            shakeOffsetY = 0
        }

        if (crashAnimProgress >= 0.8f) {
            state = GameScreenState.GAME_OVER
        }
    }

    private fun spawnBurstParticles(cx: Float, cy: Float, count: Int = 16) {
        for (i in 0 until count) {
            val p = pools.obtainParticle() ?: break
            p.isActive = true
            p.x = cx
            p.y = cy
            p.vx = ((-30..30).random()).toFloat()
            p.vy = ((-40..20).random()).toFloat()
            p.life = 0f
            p.maxLife = 0.5f
            p.colorStepIndex = 0
        }
    }

    private fun spawnMoneyRain() {
        for (i in 0 until 12) {
            val coin = pools.obtainCoin() ?: break
            coin.isActive = true
            coin.isCollected = false
            val minX = Constants.KERB_LEFT_X + 10f
            val maxX = Constants.KERB_RIGHT_X - 10f
            coin.x = minX + kotlin.random.Random.nextFloat() * (maxX - minX)
            coin.y = -((i * 15f) + 10f)
        }
    }

    private fun updateParticles(dt: Float) {
        for (p in pools.particlePool) {
            if (!p.isActive) continue
            p.life += dt
            if (p.life >= p.maxLife) {
                p.isActive = false
            } else {
                p.x += p.vx * dt
                p.y += p.vy * dt
                p.colorStepIndex = (p.life / p.maxLife * 4).toInt().coerceIn(0, 3)
            }
        }
    }

    private fun updateFloatingText(dt: Float) {
        for (ft in pools.floatingTextPool) {
            if (!ft.isActive) continue
            ft.life += dt
            if (ft.life >= ft.maxLife) {
                ft.isActive = false
            } else {
                ft.y -= 12f * dt // Rises 12px
            }
        }
    }
}

File: flutter_app/lib/game/game_engine.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';
import '../core/constants.dart';
import '../core/palette.dart';
import '../entities/game_entities.dart';
import '../entities/object_pools.dart';
import '../game/audio_director.dart';
import '../game/difficulty_manager.dart';
import '../game/spawn_director.dart';

enum GameScreenState {
  home,
  playing,
  paused,
  crashing,
  gameOver,
  leaderboard,
  achievements,
  settings,
}

class GameEngine extends ChangeNotifier {
  final AudioDirector audioDirector;

  GameEngine(this.audioDirector);

  var state = GameScreenState.home;

  final player = RickshawPlayer();
  final pools = ObjectPools();
  final spawnDirector = SpawnDirector();
  final difficultyManager = DifficultyManager();

  double survivalTimeSec = 0;
  double distanceMeters = 0;
  int moneyTaka = 0;
  int comboCount = 0;
  double comboTimerSec = 0;
  int maxCombo = 0;
  GameZone currentZone = GameZone.gulistan;

  int shakeOffsetX = 0;
  int shakeOffsetY = 0;
  double shakeTimerSec = 0;

  double crashAnimProgress = 0;
  int wipeColumnIndex = -1;

  PassengerEntity? currentPassenger;

  int lastThousandCoins = 0;
  double obstacleSpawnYTracker = 0;

  bool useBanglaDigits = true;
  bool showFpsOverlay = false;

  void startNewGame() {
    pools.resetAll();
    player.x = Constants.laneCenters[1] - (Constants.rickshawWidth / 2);
    player.lane = 1;
    player.lives = 3;
    player.invincibleTimerSec = 0;
    player.isTurbo = false;
    player.isShielded = false;
    player.isMagnetActive = false;
    player.nearMissCount = 0;

    survivalTimeSec = 0;
    distanceMeters = 0;
    moneyTaka = 0;
    lastThousandCoins = 0;
    obstacleSpawnYTracker = 0;
    spawnDirector.reset();
    comboCount = 0;
    comboTimerSec = 0;
    maxCombo = 0;
    currentZone = GameZone.gulistan;
    currentPassenger = null;
    crashAnimProgress = 0;
    wipeColumnIndex = -1;

    state = GameScreenState.playing;
    audioDirector.triggerVoice("abar_suru");
    notifyListeners();
  }

  void handleSwipe(double deltaX) {
    if (state != GameScreenState.playing) return;
    if (deltaX > 18 && player.lane < Constants.laneCount - 1) {
      final newLane = player.lane + 1;
      player.startLaneChange(newLane);
      spawnDirector.onPlayerLaneChanged(newLane);
      audioDirector.playSfx("near_miss");
    } else if (deltaX < -18 && player.lane > 0) {
      final newLane = player.lane - 1;
      player.startLaneChange(newLane);
      spawnDirector.onPlayerLaneChanged(newLane);
      audioDirector.playSfx("near_miss");
    }
  }

  void handleTapBell() {
    if (state != GameScreenState.playing) return;
    audioDirector.playSfx("bell_ting");

    for (final ped in pools.pedestrianPool) {
      if (ped.isActive && ped.y >= player.y - 40 && ped.y <= player.y + 40) {
        ped.isDodging = true;
        ped.dodgeOffset = ped.x < player.x ? -6 : 6;
      }
    }
  }

  void pauseGame() {
    if (state == GameScreenState.playing) {
      state = GameScreenState.paused;
      notifyListeners();
    }
  }

  void resumeGame() {
    if (state == GameScreenState.paused) {
      state = GameScreenState.playing;
      notifyListeners();
    }
  }

  void goToHome() {
    state = GameScreenState.home;
    notifyListeners();
  }

  void goToLeaderboard() {
    state = GameScreenState.leaderboard;
    notifyListeners();
  }

  void goToAchievements() {
    state = GameScreenState.achievements;
    notifyListeners();
  }

  void goToSettings() {
    state = GameScreenState.settings;
    notifyListeners();
  }

  void update(double dt) {
    if (state != GameScreenState.playing && state != GameScreenState.crashing) return;

    if (state == GameScreenState.crashing) {
      _updateCrashSequence(dt);
      return;
    }

    survivalTimeSec += dt;
    final tier = difficultyManager.getTierForTime(survivalTimeSec);

    final thousandCoinsCount = moneyTaka ~/ 1000;
    if (thousandCoinsCount > lastThousandCoins) {
      lastThousandCoins = thousandCoinsCount;
      addReward(0, "SPEED BOOST!");
      audioDirector.playSfx("turbo_boost");
    }

    final coinSpeedBonus = thousandCoinsCount * 14;
    var currentSpeed = tier.scrollSpeed + coinSpeedBonus;
    if (player.isTurbo) {
      currentSpeed *= Constants.turboSpeedMultiplier;
    }

    distanceMeters += currentSpeed * dt * Constants.metersPerPixel;

    final zoneIndex = ((distanceMeters / Constants.zoneIntervalMeters).toInt()) % GameZone.values.length;
    currentZone = GameZone.values[zoneIndex];

    player.update(dt);

    if (comboCount > 0) {
      comboTimerSec -= dt;
      if (comboTimerSec <= 0) {
        comboCount = 0;
      }
    }

    if (player.isMagnetActive) {
      for (final coin in pools.coinPool) {
        if (coin.isActive && !coin.isCollected) {
          final dx = player.x + 12 - coin.x;
          final dy = player.y + 16 - coin.y;
          final distSq = dx * dx + dy * dy;
          if (distSq < 48 * 48) {
            coin.x += dx / 4;
            coin.y += dy / 4;
          }
        }
      }
    }

    _updateSpawning(dt, currentSpeed, tier.maxLanesBlocked);
    _updateEntitiesAndCollisions(dt, currentSpeed);
  }

  void _updateSpawning(double dt, double currentSpeed, int maxBlockedLanes) {
    obstacleSpawnYTracker += currentSpeed * dt;

    final spawnDecision = spawnDirector.shouldSpawnObstacleRow(obstacleSpawnYTracker, currentSpeed, maxBlockedLanes);
    if (spawnDecision != null) {
      for (final blockedLane in spawnDecision.blockedLanes) {
        final obs = pools.obtainObstacle();
        if (obs != null) {
          obs.isActive = true;
          obs.x = Constants.laneCenters[blockedLane] - 8;
          obs.y = -30;
          obs.type = spawnDecision.obstacleType;
        }
      }
    }

    if (Random().nextDouble() < 0.02) {
      final pattern = spawnDirector.generateCoinPattern(-20);
      for (final info in pattern) {
        final coin = pools.obtainCoin();
        if (coin != null) {
          coin.isActive = true;
          coin.isCollected = false;
          coin.x = info.x;
          coin.y = info.y;
        }
      }
    }

    final pupType = spawnDirector.checkPowerUpSpawn();
    if (pupType != null) {
      final pup = pools.obtainPowerUp();
      if (pup != null) {
        pup.isActive = true;
        pup.type = pupType;
        final lane = Random().nextInt(Constants.laneCount);
        pup.x = Constants.laneCenters[lane] - 8;
        pup.y = -20;
      }
    }

    if (Random().nextDouble() < 0.015) {
      final ped = pools.obtainPedestrian();
      if (ped != null) {
        ped.isActive = true;
        ped.isDodging = false;
        ped.dodgeOffset = 0;
        final lane = Random().nextInt(Constants.laneCount);
        ped.x = Constants.laneCenters[lane] - 8;
        ped.y = -24;
      }
    }

    if (currentPassenger == null && Random().nextDouble() < 0.008) {
      final pass = pools.obtainPassenger();
      if (pass != null) {
        pass.isActive = true;
        pass.isLeftKerb = Random().nextBool();
        pass.x = pass.isLeftKerb ? 10 : 154;
        pass.y = -24;
        pass.typeIndex = Random().nextInt(6);
        pass.isPickedUp = false;
        pass.dropoffTargetMeter = distanceMeters.toInt() + Random().nextInt(500) + 400;
      }
    }
  }

  void _updateEntitiesAndCollisions(double dt, double currentSpeed) {
    final playerHitbox = player.getHitbox();

    for (final obs in pools.obstaclePool) {
      if (!obs.isActive) continue;
      obs.y += currentSpeed * dt;

      if (obs.getHitbox().intersects(playerHitbox)) {
        if (player.isTurbo) {
          obs.isActive = false;
          _spawnBurstParticles(obs.x + 8, obs.y + 8);
          addReward(10, "SMASH!");
        } else if (player.isShielded) {
          player.isShielded = false;
          obs.isActive = false;
          audioDirector.playSfx("near_miss");
          _spawnBurstParticles(obs.x + 8, obs.y + 8);
        } else if (player.invincibleTimerSec > 0) {
          // invincible
        } else {
          player.lives--;
          obs.isActive = false;
          _spawnBurstParticles(obs.x + 8, obs.y + 8, count: 16);
          shakeTimerSec = 0.35;
          audioDirector.playSfx("crash_boom");

          if (player.lives > 0) {
            player.invincibleTimerSec = 1.8;
            addReward(0, "-1 LIFE!");
            audioDirector.triggerVoice("are_baba");
          } else {
            _triggerCrash();
            return;
          }
        }
      }

      if (obs.y > Constants.virtualHeight + 32) obs.isActive = false;
    }

    for (final coin in pools.coinPool) {
      if (!coin.isActive || coin.isCollected) continue;
      coin.y += currentSpeed * dt;

      if (coin.getHitbox().intersects(playerHitbox)) {
        coin.isCollected = true;
        coin.isActive = false;
        addReward(20, "Taka20");
        audioDirector.playSfx("coin_ching");
      }

      if (coin.y > Constants.virtualHeight + 16) coin.isActive = false;
    }

    for (final pup in pools.powerUpPool) {
      if (!pup.isActive) continue;
      pup.y += currentSpeed * dt;

      if (pup.getHitbox().intersects(playerHitbox)) {
        pup.isActive = false;
        audioDirector.playSfx("turbo_boost");
        switch (pup.type) {
          case PowerUpType.turbo:
            player.isTurbo = true;
            player.turboTimeSec = pup.type.durationSec;
            audioDirector.triggerVoice("bachao");
          case PowerUpType.moneyMagnet:
            player.isMagnetActive = true;
            player.magnetTimeSec = pup.type.durationSec;
          case PowerUpType.shield:
            player.isShielded = true;
          case PowerUpType.moneyRain:
            _spawnMoneyRain();
        }
      }

      if (pup.y > Constants.virtualHeight + 16) pup.isActive = false;
    }

    for (final pass in pools.passengerPool) {
      if (!pass.isActive) continue;
      pass.y += currentSpeed * dt;

      if (!pass.isPickedUp && pass.getHitbox().intersects(playerHitbox)) {
        pass.isPickedUp = true;
        pass.isActive = false;
        currentPassenger = pass;
        addReward(50, "PICKUP!");
        audioDirector.playSfx("pickup_ting");
        audioDirector.triggerVoice("vada_ache");
      }

      if (pass.y > Constants.virtualHeight + 24) pass.isActive = false;
    }

    if (currentPassenger != null && distanceMeters >= currentPassenger!.dropoffTargetMeter) {
      addReward(200, "DROPOFF!");
      audioDirector.playSfx("dropoff_chime");
      currentPassenger = null;
    }

    for (final ped in pools.pedestrianPool) {
      if (!ped.isActive) continue;
      ped.y += currentSpeed * dt;

      final pedHitbox = ped.getHitbox();
      if (pedHitbox.intersects(playerHitbox)) {
        ped.isActive = false;
        addReward(100, "+100 PTS!");
        audioDirector.playSfx("coin_ching");
        audioDirector.triggerVoice("mama_side_den");
        _spawnBurstParticles(ped.x + 8, ped.y + 12, count: 12);
      } else {
        final distY = (ped.y - player.y).abs();
        if (distY < 6 && !ped.isDodging) {
          ped.isDodging = true;
          ped.dodgeOffset = ped.x < player.x ? -8 : 8;
          player.nearMissCount++;
          addReward(10, "NEAR MISS!");
          audioDirector.playSfx("near_miss");
          audioDirector.triggerVoice("mama_side_den");
        }
      }

      if (ped.y > Constants.virtualHeight + 24) ped.isActive = false;
    }

    _updateParticles(dt);
    _updateFloatingText(dt);
  }

  void addReward(int baseTaka, String label) {
    comboCount++;
    comboTimerSec = 3.0;
    if (comboCount > maxCombo) maxCombo = comboCount;

    final multiplier = comboCount >= 20
        ? 5
        : comboCount >= 10
            ? 3
            : comboCount >= 5
                ? 2
                : 1;

    final total = baseTaka * multiplier;
    moneyTaka += total;

    final ft = pools.obtainFloatingText();
    if (ft != null) {
      ft.isActive = true;
      ft.text = multiplier > 1 ? "$label x$multiplier" : label;
      ft.x = player.x;
      ft.y = player.y - 10;
      ft.life = 0;
      ft.maxLife = 0.6;
      final color = multiplier == 5
          ? Palette.gold
          : multiplier == 3
              ? Palette.amber
              : multiplier == 2
                  ? Palette.orange
                  : Palette.cream;
      ft.colorArgb = color.toARGB32();
    }
  }

  void _triggerCrash() {
    player.lives = 0;
    state = GameScreenState.crashing;
    audioDirector.playSfx("crash_boom");
    audioDirector.triggerVoice("are_baba");

    _spawnBurstParticles(player.x + 16, player.y + 20, count: 24);
    shakeTimerSec = 0.35;
    notifyListeners();
  }

  void _updateCrashSequence(double dt) {
    crashAnimProgress += dt;
    if (shakeTimerSec > 0) {
      shakeTimerSec -= dt;
      shakeOffsetX = Random().nextInt(7) - 3;
      shakeOffsetY = Random().nextInt(7) - 3;
    } else {
      shakeOffsetX = 0;
      shakeOffsetY = 0;
    }

    if (crashAnimProgress >= 0.8) {
      state = GameScreenState.gameOver;
      notifyListeners();
    }
  }

  void _spawnBurstParticles(double cx, double cy, {int count = 16}) {
    for (var i = 0; i < count; i++) {
      final p = pools.obtainParticle();
      if (p == null) break;
      p.isActive = true;
      p.x = cx;
      p.y = cy;
      p.vx = (Random().nextInt(61) - 30).toDouble();
      p.vy = (Random().nextInt(61) - 40).toDouble();
      p.life = 0;
      p.maxLife = 0.5;
      p.colorStepIndex = 0;
    }
  }

  void _spawnMoneyRain() {
    for (var i = 0; i < 12; i++) {
      final coin = pools.obtainCoin();
      if (coin == null) break;
      coin.isActive = true;
      coin.isCollected = false;
      final minX = Constants.kerbLeftX + 10;
      final maxX = Constants.kerbRightX - 10;
      coin.x = minX + Random().nextDouble() * (maxX - minX);
      coin.y = -((i * 15) + 10);
    }
  }

  void _updateParticles(double dt) {
    for (final p in pools.particlePool) {
      if (!p.isActive) continue;
      p.life += dt;
      if (p.life >= p.maxLife) {
        p.isActive = false;
      } else {
        p.x += p.vx * dt;
        p.y += p.vy * dt;
        p.colorStepIndex = (p.life / p.maxLife * 4).toInt().clamp(0, 3);
      }
    }
  }

  void _updateFloatingText(double dt) {
    for (final ft in pools.floatingTextPool) {
      if (!ft.isActive) continue;
      ft.life += dt;
      if (ft.life >= ft.maxLife) {
        ft.isActive = false;
      } else {
        ft.y -= 12 * dt;
      }
    }
  }
}
